File: src/main/java/org/radarcns/monitor/CombinedKafkaMonitor.java
/*
 * Copyright 2017 Kings College London and The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.monitor;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Runs multiple monitors, each in its own thread.
 */
public class CombinedKafkaMonitor implements KafkaMonitor {
    private static final Logger logger = LoggerFactory.getLogger(CombinedKafkaMonitor.class);

    private final List<KafkaMonitor> monitors;
    private final AtomicBoolean done;

    private ExecutorService executor;
    private IOException ioException;
    private InterruptedException interruptedException;

    public CombinedKafkaMonitor(Collection<KafkaMonitor> monitors) {
        if (monitors == null || monitors.isEmpty()) {
            throw new IllegalArgumentException("Monitor collection may not be empty");
        }
        this.monitors = new ArrayList<>(monitors);
        this.done = new AtomicBoolean(false);
        this.executor = null;
        this.ioException = null;
        this.interruptedException = null;
    }

    @Override
    public long getPollTimeout() {
        return monitors.get(0).getPollTimeout();
    }

    @Override
    public void setPollTimeout(long pollTimeout) {
        for (KafkaMonitor monitor : monitors) {
            monitor.setPollTimeout(pollTimeout);
        }
    }

    @Override
    public void start() throws IOException, InterruptedException {
        synchronized (this) {
            if (executor != null) {
                throw new IllegalStateException("Cannot start monitor twice");
            }
            executor = Executors.newFixedThreadPool(monitors.size());
        }

        for (KafkaMonitor monitor : monitors) {
            executor.submit(() -> {
                try {
                    monitor.start();
                } catch (IOException ex) {
                    setIoException(ex);
                } catch (InterruptedException ex) {
                    setInterruptedException(ex);
                }
            });
        }

        executor.shutdown();
        executor.awaitTermination(366_000, TimeUnit.DAYS); // > 1000 years...

        if (getIoException() != null) {
            throw getIoException();
        }
        if (getInterruptedException() != null) {
            throw getInterruptedException();
        }
    }

    private synchronized IOException getIoException() {
        return ioException;
    }

    private synchronized void setIoException(IOException ioException) {
        this.ioException = ioException;
        initiateShutdownIgnoreException();
    }

    private synchronized InterruptedException getInterruptedException() {
        return interruptedException;
    }

    private synchronized void setInterruptedException(InterruptedException interruptedException) {
        this.interruptedException = interruptedException;
        initiateShutdownIgnoreException();
    }

    private void initiateShutdownIgnoreException() {
        try {
            initiateShutdown();
        } catch (InterruptedException ex) {
            logger.info("Ignoring additional InterruptedException", ex);
        } catch (IOException ex) {
            logger.info("Ignoring additional IOException", ex);
        }
    }

    private void initiateShutdown() throws IOException, InterruptedException {
        if (!done.getAndSet(true)) {
            for (KafkaMonitor monitor : monitors) {
                monitor.shutdown();
            }
        }
    }

    @Override
    public void shutdown() throws IOException, InterruptedException {
        synchronized (this) {
            if (executor == null) {
                return;
            }
        }
        initiateShutdown();
        executor.awaitTermination(30, TimeUnit.SECONDS);
        executor.shutdownNow();
    }

    @Override
    public boolean isShutdown() {
        return done.get();
    }
}

File: src/main/kotlin/org/radarbase/monitor/CombinedKafkaMonitor.kt
/*
 * Copyright 2017 King's College London and The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.monitor

import org.slf4j.LoggerFactory
import java.io.IOException
import java.time.Duration
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.stream.Collectors
import java.util.stream.Stream

/**
 * Runs multiple monitors, each in its own thread.
 */
class CombinedKafkaMonitor(monitors: Stream<KafkaMonitor>) : KafkaMonitor {
    private val monitors: List<KafkaMonitor> =
        Objects.requireNonNull(monitors).filter(Objects::nonNull).collect(Collectors.toList())

    private val done = AtomicBoolean(false)
    private var executor: ExecutorService? = null
    private var ioException: IOException? = null
    private var interruptedException: InterruptedException? = null

    init {
        require(this.monitors.isNotEmpty()) { "Monitor collection may not be empty" }
    }

    override val isShutdown: Boolean
        get() = done.get()

    override var pollTimeout: Duration
        get() = monitors[0].pollTimeout
        set(pollTimeout) {
            for (monitor in monitors) {
                monitor.pollTimeout = pollTimeout
            }
        }

    override fun start() {
        synchronized(this) {
            if (executor != null) {
                throw IllegalStateException("Cannot start monitor twice")
            }
            executor = Executors.newFixedThreadPool(monitors.size)
        }

        val currentExecutor = executor!!
        for (monitor in monitors) {
            currentExecutor.submit {
                try {
                    monitor.start()
                } catch (ex: IOException) {
                    setIoException(ex)
                } catch (ex: InterruptedException) {
                    setInterruptedException(ex)
                }
            }
        }

        currentExecutor.shutdown()
        currentExecutor.awaitTermination(366_000, TimeUnit.DAYS) // > 1000 years...

        getIoException()?.let { throw it }
        getInterruptedException()?.let { throw it }
    }

    @Synchronized
    private fun getIoException(): IOException? = ioException

    @Synchronized
    private fun setIoException(ioException: IOException) {
        this.ioException = ioException
        initiateShutdownIgnoreException()
    }

    @Synchronized
    private fun getInterruptedException(): InterruptedException? = interruptedException

    @Synchronized
    private fun setInterruptedException(interruptedException: InterruptedException) {
        this.interruptedException = interruptedException
        initiateShutdownIgnoreException()
    }

    private fun initiateShutdownIgnoreException() {
        try {
            initiateShutdown()
        } catch (ex: InterruptedException) {
            logger.info("Ignoring additional InterruptedException", ex)
        } catch (ex: IOException) {
            logger.info("Ignoring additional IOException", ex)
        }
    }

    private fun initiateShutdown() {
        if (!done.getAndSet(true)) {
            for (monitor in monitors) {
                try {
                    monitor.shutdown()
                } catch (ex: Exception) {
                    logger.warn("Failed to shutdown monitor ${monitor.javaClass.simpleName}", ex)
                }
            }
        }
    }

    override fun shutdown() {
        synchronized(this) {
            if (executor == null) {
                return
            }
        }
        initiateShutdown()
        executor?.let {
            it.awaitTermination(30, TimeUnit.SECONDS)
            it.shutdownNow()
        }
    }

    fun getMonitors(): List<KafkaMonitor> = ArrayList(monitors)

    companion object {
        private val logger = LoggerFactory.getLogger(CombinedKafkaMonitor::class.java)
    }
}
